sdks/java/core/src/main/java/org/apache/beam/sdk/util/ComparableThrowable.java

```diff
@@ -0,0 +1,112 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import javax.annotation.Nullable;
+
+/**
+ * A wrapper around a {@link Throwable} for use with coders.
+ *
+ * <p>The {@link Throwable} class is serializable. However, it, along with man child classes like
+ * {@link Exception}, does not override {@link Object#equals(Object)}. As a result, t1.equals(t2)
+ * will return false if t2 is a deserialized instance of serialized t1. This makes it appear as if
+ * coders are mutating the value, which can lead to things like log spam. This wrapper overrides
+ * {@link Object#equals(Object)} to get around this issue. The hash code remains the same as the
+ * underlying {@link Throwable}.
+ *
+ * <p>The equality comparison is transitive, meaning that for two {@link Throwable}s t1 and t2, the
+ * causes must be equal. This can occur either because they are both null or because they are both
+ * equal after being wrapped.
+ *
+ * <p>Note that this is simply a best effort based on properties like instance type, stack trace,
+ * message, and cause. It cannot guarantee that the state that led to the exception is the same,
+ * unless it is fully captured in the message, nor can it differentiate between two exceptions
+ * thrown at different times or in different processes. For this reason, this is not suitable for a
+ * general-purpose {@link Throwable} equality comparison. It merely exists to try to avoid false
+ * positives in a mutation check for coders.
+ *
+ * <p>Due to the above, this does not support comparison to a raw {@link Throwable}.
+ */
+public class ComparableThrowable implements Serializable {
+  private Throwable throwable;
+
+  private ComparableThrowable() {
+    // Can't set this to null without adding a pointless @Nullable annotation to the field. It also
+    // needs to be set from the constructor to avoid a checkstyle violation.
+    this.throwable = new Throwable();
+  }
+
+  /** Wraps {@code throwable} and returns the result. */
+  public static ComparableThrowable forThrowable(Throwable throwable) {
+    ComparableThrowable comparable = new ComparableThrowable();
+    comparable.throwable = throwable;
+    return comparable;
+  }
+
+  /** Returns the underlying {@link Throwable}. */
+  public Throwable throwable() {
+    return throwable;
+  }
+
+  @Override
+  public int hashCode() {
+    return throwable.hashCode();
+  }
+
+  @Override
+  public boolean equals(@Nullable Object obj) {
+    if (!(obj instanceof ComparableThrowable)) {
+      return false;
+    }
+    Throwable other = ((ComparableThrowable) obj).throwable;
+
+    boolean currentLevelEqual =
+        throwable.getClass().isInstance(other)
```

Comment:
So just `return throwable.getClass().equals(other.getClass())` would be enough? I guess it could work, since this is just for after a built-in type has been deserialized. (And if Java is messing up serialization/deserialization of built-in types, we have far greater issues.)

Reply:
Yea